File: 2025/kotlin/src/main/kotlin/Day02.kt
package de.rubixdev

fun day2(): Day = sequence {
    val example = "11-22,95-115,998-1012,1188511880-1188511890,222220-222224," +
        "1698522-1698528,446443-446449,38593856-38593862,565653-565659," +
        "824824821-824824827,2121212118-2121212124"
//    val input = example
    val input = getInput(2)
        .split(',')
        .map { range -> range.split('-').map { it.toLong() }.toPair() }
        .map { (start, end) -> start..end }

    yield(Unit)
    yield(part1(input))
    yield(part2(input))
}

private fun part(input: List<LongRange>, regex: Regex): Long = input
    .asSequence()
    .flatten()
    .filter { regex.matches(it.toString()) }
    .sum()

private fun part1(input: List<LongRange>) = part(input, "^([0-9]+)\\1$".toRegex())
private fun part2(input: List<LongRange>) = part(input, "^([0-9]+)\\1+$".toRegex())

File: 2025/scala/src/main/scala/de/rubixdev/aoc/Day02.scala
package de.rubixdev.aoc

import scala.util.matching.Regex

object Day02 extends Day {
  override def run(rawInput: String): Unit = {
    val input = rawInput.trim
      .split(',')
      .map(_.split('-').map(_.toLong))
      .map { nums => nums(0) to nums(1) }
      .toList

    println("Part 1: " + part1(input))
    println("Part 2: " + part2(input))
  }

  private def part(input: List[Iterable[Long]], regex: Regex) =
    input.view.flatten.filter { num => regex.matches(num.toString) }.sum

  private def part1(input: List[Iterable[Long]]) =
    part(input, Regex("^([0-9]+)\\1$"))
  private def part2(input: List[Iterable[Long]]) =
    part(input, Regex("^([0-9]+)\\1+$"))
}
